Sources/NIO/CircularBuffer.swift

```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
         }
     }
 
+    /// Replaces the specified subrange of elements with the given collection.
+    ///
+    /// - Parameter subrange:
+    /// The subrange of the collection to replace. The bounds of the range must be valid indices of the collection.
+    ///
+    /// - Parameter newElements:
+    /// The new elements to add to the collection.
+    ///
+    /// *O(n)* where _n_ is the lenght of the new elements collection if the subrange equals to _n_
+    ///
+    /// *O(m)* where _m_ is the combined length of the collection and _newElements_
+    public mutating func replaceSubrange(_ subrange: RangeType<Int>, with newElements: ContiguousArray<E>) {
+        guard subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex else {
```

Comment:
`endIndex` is actually one step *past* the end of the array, so the upper bound should be closed (e.g. `subrange.upperBound < self.endIndex`.